packages/zone.js/lib/awaiter/zone.native.awaiter.ts

```
@@ -0,0 +1,26 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+Zone.__load_patch('native.awaiter', () => {
+  Zone.__awaiter = async function<T>(thisArg: any, _arguments: any, generator: () => Generator<T>) {
+    const zone = Zone.current;
+    const gen = generator.apply(thisArg, _arguments);
+    let res = gen.next();
+    const args = [await res.value];
+    while (!res.done) {
+      try {
+        res = zone.run(gen.next, gen, args, 'native await');
+        args[0] = await res.value;
+      } catch (error) {
+        zone.runGuarded(() => {
```

Why use `runGuarded()` here? What will happen if the `gen.throw()` call also throws?

---

Using `runGuarded()` is to trigger `zoneSpec.onHandleError` hook to let zone.js handle the error. And what do you mean `gen.throw()` also throw? `gen.throw()` will throw an error and caught by `zone.runGuarded()`.

---

And what would `zoneSpec.onHandleError()` do with the error? Would it rethrow it?

---

`zoneSpec.onHandleError()` by default will rethrow the error, but it also supports not rethrow and swallow the error by `return false` in the `onHandleError()` hook.

---

I see, thx!
I am still not 100% sure why we call `runGuarded()` here 😅
What does `zoneSpec.onHandleError()` do? When/How would it be called in the equivalent Promise-based code?

---

A similar question... if we need `runGuarded()` here, why not on line 17 above?

---

What would happen if we had a function that looked like:

```ts
async function foo() {
  try {
    await throw 'error 1';
  } catch (e1) {
    try {
      await throw 'error 2';
    } catch (e2) {
      doSomething();
    }
  }
}
```

Would the 2nd `try-catch` in the async function ever get executed?